src/main/java/io/reactivex/internal/schedulers/ScheduledRunnable.java

```diff
@@ -99,23 +108,24 @@ public void setFuture(Future<?> f) {
     public void dispose() {
         for (;;) {
             Object o = get(FUTURE_INDEX);
-            if (o == DONE || o == DISPOSED) {
+            if (o == DONE || o == SYNC_DISPOSED || o == ASYNC_DISPOSED) {
                 break;
             }
-            if (compareAndSet(FUTURE_INDEX, o, DISPOSED)) {
+            boolean async = get(THREAD_INDEX) != Thread.currentThread();
+            if (compareAndSet(FUTURE_INDEX, o, async ? ASYNC_DISPOSED : SYNC_DISPOSED)) {
```

Comment:
Hm, what worries me here is that we base `boolean async` on `get(THREAD_INDEX) != Thread.currentThread()`, but in `finally` of `run()` we set `Thread` to null which always gonna give `async == false` 

https://github.com/ReactiveX/RxJava/blob/b99efbfb71d3a76befccd4e36324cefadeefd65a/src/main/java/io/reactivex/internal/schedulers/ScheduledRunnable.java#L67

Yes, it's `lazySet()`, but aren't we creating unwanted race condition here between `run()` and `dispose()`?

The worst scenario is that if `dispose()` indeed was called from another thread but `lazySet()` memory write becomes visible to it, we'll have `async == false` which is actually wrong in this scenario…

Reply:
Cancelling from the current thread could only happen while the `run` callback is executing. Any other cancellation that comes after this can only be asynchronous. It doesn't matter what value is in `THREAD_INDEX` at this time because that will not be the same thread as the caller's. 

Since we are running with single-threaded thread pools, any subsequent task on the same pool having the same Thread will find the references marked as `DONE` and not cancel the previous task. This is guaranteed by the sequential task processing of the thread pool.

Reply:
>It doesn't matter what value is in THREAD_INDEX at this time because that will not be the same thread as the caller's.

Aha, I see now, 👍